### ISSUE TITLE
Unity iOS: Errors while creating transport!.

### ISSUE THREAD

**Issue by eAi** [ably/ably-dotnet#764]
When running on iOS with Unity I get error 80019.

Due to the aforementioned bug #762 this doesn’t provide much information. I dug deeper into this by setting the debug level to Debug and providing a custom Unity logger. This shows this error is occurring:

    Errors while creating transport!.
    
    IO.Ably.InternalLogger:Message(LogLevel, String, Object[])
    IO.Ably.Transport.<CreateTransport>d__33:MoveNext()
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
    IO.Ably.Transport.<CreateTransportParameters>d__43:MoveNext()
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
    IO.Ably.Transport.<Create>d__61:MoveNext()
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
    IO.Ably.<GetCurrentValidTokenAndRenewIfNecessaryAsync>d__56:MoveNext()
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
    IO.Ably.<RenewToken>d__57:MoveNext()
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetException(Exception)
    IO.Ably.<RequestTokenAsync>d__58:MoveNext()
    System.Threading.ThreadPoolWorkQueue:Dispatch() Unfortunately this doesn’t provide much information - there’s no exception provided after the message as it appears it should be - and the way the transports work is hard to follow through. I’ve tried setting the stripping level to Low in case that was the cause but that makes no difference. I’ll see if I can reproduce in a small project tomorrow.

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/SDK-1148) by [Unito](https://www.unito.io)

**Comment by marto83** [comment]
While the transport is created it will try to authenticate, which explains the transport error. Did you get the full debug output? If you have an AuthUrl specified you will get the full request and response in the logs. 

Things to try:
1. Test with an application key just to rule out anything weird happening with Unity
2. Try with the custom ably dlls attached. I've compiled it after adding the exception to the error info returned so it gives you more information. 
3. Attach the full debug log and I can try and work out what is happening. 

[ably-custom.zip](https://github.com/ably/ably-dotnet/files/7020481/ably-custom.zip)

**Comment by eAi** [comment]
Hi @marto83 -  thanks for that. I've tried with just an application key and those DLLs. Unfortunately those DLLs cause a new exception:

```
NullReferenceException: Object reference not set to an instance of an object.

IO.Ably.IoC.get_MobileDevice () (at <00000000000000000000000000000000>:0)
IO.Ably.Push.PushRealtime..ctor (IO.Ably.AblyRest restClient, IO.Ably.ILogger logger) (at <00000000000000000000000000000000>:0)
IO.Ably.AblyRealtime..ctor (IO.Ably.ClientOptions options, System.Func`2[T,TResult] createRestFunc) (at <00000000000000000000000000000000>:0)
```

Presumably this on this line: https://github.com/ably/ably-dotnet/blob/f2106ee6c68681a1d8533767b665ea7bbae9acb8/src/IO.Ably.Shared/IoC.cs#L46

**Comment by eAi** [comment]
Here's a full log and the test code that reproduces this error: https://gist.github.com/eAi/0a9a32af2d91953d43be72ad73287858

**Comment by marto83** [comment]
@eAi Sorry about that. My bad, I've given you a dll with an outstanding bug. I'll send another one shortly.

**Comment by marto83** [comment]
Try with this version. 

I've also noticed that you wait to create the channel after connected. You don't need to do that as the library will internally manage it for you and it will attach the channel once the connection is established. Let me know how it goes. I'll keep an eye for notifications.

[ably-custom-v2.zip](https://github.com/ably/ably-dotnet/files/7021399/ably-custom-v2.zip)

**Comment by eAi** [comment]
@marto83 I'm getting exactly the same error as in the [gist above](https://gist.github.com/eAi/0a9a32af2d91953d43be72ad73287858) with the new DLLs in my simple test project.

Thanks for letting me know about the channels - I assumed that might be the case, but did that to be safe - good to know it can be simplified!

**Comment by marto83** [comment]
That's definitely odd. I've put this part inside an try catch block just in case. Could be it a cached dll?
![image](https://user-images.githubusercontent.com/150508/130236105-207889b2-7452-40e6-a668-4713bba938eb.png)

In case it is something I've done wrong. I did a clean and rebuild and here are the dlls again
[ably-custom-v3.zip](https://github.com/ably/ably-dotnet/files/7021589/ably-custom-v3.zip)

**Comment by eAi** [comment]
@marto83 This is now working - somewhat! Ably seems to be able to startup correctly and connect - and receive messages. Unfortunately, it seems as though that on device (not in the Unity editor) the messages from the Ably servers aren't being read correctly?

On Device I see this:
```
02:50:06.622 Websocket message received. Raw: {"action":15,"id":"l-GZc2q3Qv:0","connectionId":"l-GZc2q3Qv","connectionSerial":2,"channel":"user_611d519b3255ce000941423a","channelSerial":"108JY91sQB1hjD82984091:4","timestamp":1629471006397,"messages":[{"data":<snip>,"name":"notification"}]}
...
ABLY message: Id: l-GZc2q3Qv:0:0, Name: , Data: <snip>, Encoding: , Timestamp: 
```

In the editor I see this:
```
02:48:38.021 Websocket message received. Raw: {"action":15,"id":"d8TEQvBeIa:4","connectionId":"d8TEQvBeIa","connectionSerial":1,"channel":"user_611e1f0515e86400090964e1","channelSerial":"108tszqAwB1heK07131977:1","timestamp":1629470917980,"messages":[{"data":<snip>,"name":"notification"}]}

ABLY message: Id: d8TEQvBeIa:4:0, Name: notification, Data: <snip>, Encoding: , Timestamp: 20/08/2021 14:48:37 +00:00
```

Notice how the editor version has parsed the name 'notification' and has the timestamp, while the on device one is missing both of these. This means that the messages aren't received by our code on device.

If need be I can provide a full log if you let me know an email address.

**Comment by marto83** [comment]
@eAi it will be best if you create a support issue and include the log in the issue. This way @tomkirbygreen and I will be able to both see it. Let me know the issue number once you have created it.

**Comment by eAi** [comment]
I've emailed it to support - this doesn't appear to have created a ticket (at least not one that I can see externally), but it's on a reply to an email sent to Mark Cox-Smith - I've asked him to send it to the two of you.

**Comment by marto83** [comment]
@eAi Thinking about what could be going wrong ... I feel the issue comes from the json deserialization. It seems that the payload is deserialised correctly but not the Ably message object. 

Until we get it resolved you can work around the issues by dedicating a channel to notifications and completely ignoring the message name. Having said that, it's basic functionality to include the name so it should work. I'll let you know if I see anything in the logs. 
Btw I'm away next week so I might not be able to reply but will message Mark to get somebody else to look into it.

**Comment by eAi** [comment]
@marto83 That was my conclusion too so I did the same thing you suggested which solves the problem, notifications are now working on device - so we're all good for now, but clearly something isn't working quite correctly. 

Thank you for all your help with this - have a good week away!

**Comment by eAi** [comment]
I think that the underlying issue that started this whole thread off - and the issue with the broken deserialisation is all caused by stripping. Adding IO.Ably to our link.xml seems to resolve both these issues.

I suggest you include this in your unitypackage (you can have multiple link.xml files, Unity will merge them together).

```
<linker>
    <assembly fullname="IO.Ably" preserve="all" />
</linker>
```

There is probably a more subtle way to do this - perhaps preserving the whole assembly is unnecessary and a few necessary types required for deserialisation can be preserved instead.

**Comment by QuintinWillison** [comment]
I'm assuming this was fixed in https://github.com/ably/ably-dotnet/pull/1119 as this issue has seen no activity in over a year.
